lite2/client.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	var err error
+	switch c.mode {
+	case sequential:
+		err = c.sequence(newHeader, newVals, now)
+	case skipping:
+		err = c.bisection(c.trustedHeader, c.trustedNextVals, newHeader, newVals, now)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Update trusted header and vals.
+	nextVals, err := c.primary.ValidatorSet(newHeader.Height + 1)
+	if err != nil {
+		return err
+	}
+	return c.updateTrustedHeaderAndVals(newHeader, nextVals)
+}
+
+func (c *Client) sequence(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	// 1) Verify any intermediate headers.
+	var (
+		interimHeader *types.SignedHeader
+		nextVals      *types.ValidatorSet
+		err           error
+	)
+	for height := c.trustedHeader.Height + 1; height < newHeader.Height; height++ {
+		interimHeader, err = c.primary.SignedHeader(height)
+		if err != nil {
+			return errors.Wrapf(err, "failed to obtain the header #%d", height)
+		}
+
+		err = Verify(c.chainID, c.trustedHeader, c.trustedNextVals, interimHeader, c.trustedNextVals, c.trustingPeriod, now, c.trustLevel)
+		if err != nil {
+			return errors.Wrapf(err, "failed to verify the header #%d", height)
+		}
+
+		// Update trusted header and vals.
+		if height == newHeader.Height-1 {
+			nextVals = newVals
+		} else {
+			nextVals, err = c.primary.ValidatorSet(height + 1)
+			if err != nil {
+				return errors.Wrapf(err, "failed to obtain the vals #%d", height+1)
+			}
+		}
+		if !bytes.Equal(interimHeader.NextValidatorsHash, nextVals.Hash()) {
+			return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
+				interimHeader.NextValidatorsHash,
+				nextVals.Hash(),
+				height)
+		}
+		err = c.updateTrustedHeaderAndVals(interimHeader, nextVals)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update trusted state #%d", height)
+		}
+	}
+
+	// 2) Verify the new header.
+	return Verify(c.chainID, c.trustedHeader, c.trustedNextVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+}
+
+func (c *Client) bisection(
+	lastHeader *types.SignedHeader,
+	lastVals *types.ValidatorSet,
+	newHeader *types.SignedHeader,
+	newVals *types.ValidatorSet,
+	now time.Time) error {
+
+	err := Verify(c.chainID, lastHeader, lastVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+	switch err.(type) {
+	case nil:
+		return nil
+	case types.ErrTooMuchChange:
+		// continue bisection
+	default:
+		return errors.Wrapf(err, "failed to verify the header #%d ", newHeader.Height)
+	}
+
+	if newHeader.Height == c.trustedHeader.Height+1 {
+		// TODO: submit evidence here
+		return errors.Errorf("adjacent headers (#%d and #%d) that are not matching", lastHeader.Height, newHeader.Height)
+	}
+
+	pivot := (c.trustedHeader.Height + newHeader.Header.Height) / 2
+	pivotHeader, pivotVals, err := c.fetchHeaderAndValsAtHeight(pivot)
+	if err != nil {
+		return err
+	}
+
+	if err := c.bisection(lastHeader, lastVals, pivotHeader, pivotVals, now); err == nil {
```

Comment:
This is a confusing pattern and seems unlikely to be optimised by the compiler. Can we just do a standard tail call?

Reply:
https://stackoverflow.com/a/12116644

afaik Go only does tail call optimization in rare cases

Reply:
see 11e1e9015

Reply:
I still think this control flow could be clearer. Up to you though.

Reply:
> I still think this control flow could be clearer.

if you think it can be improved, please do so! you can commit to this branch

Reply:
It's better now!